lib/base/io-engine.hpp

```diff
@@ -167,47 +168,34 @@ class AsioConditionVariable
  */
 class Timeout : public SharedObject
 {
+private:
+	boost::asio::deadline_timer m_Timer;
+
 public:
 	DECLARE_PTR_TYPEDEFS(Timeout);
 
-	template<class Executor, class TimeoutFromNow, class OnTimeout>
-	Timeout(boost::asio::io_context& io, Executor& executor, TimeoutFromNow timeoutFromNow, OnTimeout onTimeout)
-		: m_Timer(io)
+	/**
+	 * Schedules onTimeout to be run after timeoutFromNow on strand.
+	 */
```

Comment:
```suggestion
	/**
	 * Starts an asynchronous timer that expires after the specified `timeoutFromNow` and triggers the `onTimeout` callback.
	 *
	 * Note that this does not wait for the timer to fire, but will always return immediately, thus you are
	 * responsible for keeping the Timeout object in scope for the desired duration. Otherwise, when the Timeout
	 * instance goes out of scope, it won't wait for the internal timer to expire, but will destroy it and cancel
	 * the async_wait operation unless the timer has already expired and our handler is waiting/queued to be invoked
	 * on the provided strand.
	 */
```

Reply:
My doc commit is still on top, so feel free to `git reset --hard HEAD^` and put this/similar in the class doc if you prefer your version to "my" new one. However:

> cancel the async_wait operation unless the timer has already expired and our handler is waiting/queued to be invoked

Not really. The unit tests base_io_engine/timeout_due_* even give the timer enough time to expire and wait/enqueue(?) for invocation. But then it's cancelled and not run.

Reply:
> Not really. The unit tests base_io_engine/timeout_due_* even give the timer enough time to expire and wait/enqueue(?) for invocation. But then it's cancelled and not run.

That's basically a quote from the asio documentation for `cancel()`.
<img width="584" alt="Bildschirmfoto 2024-11-29 um 12 34 38" src="https://github.com/user-attachments/assets/c10a0888-6f22-4ef5-885b-642efd17dac8">

Ah, and regarding your unit tests, add this right before the `io.run()` call and see what happens :).
```cxx
std::thread io_thread([&]() { io.run(); });
```